include/dirac_quda.h

```diff
@@ -1878,29 +1891,63 @@ namespace quda {
     virtual ~DiracMatrix() { }
 
     virtual void operator()(ColorSpinorField &out, const ColorSpinorField &in) const = 0;
-    virtual void operator()(ColorSpinorField &out, const ColorSpinorField &in,
-			    ColorSpinorField &tmp) const = 0;
-    virtual void operator()(ColorSpinorField &out, const ColorSpinorField &in,
-			    ColorSpinorField &Tmp1, ColorSpinorField &Tmp2) const = 0;
-
+    virtual void operator()(ColorSpinorField &out, const ColorSpinorField &in, ColorSpinorField &tmp) const = 0;
+    virtual void operator()(ColorSpinorField &out, const ColorSpinorField &in, ColorSpinorField &Tmp1,
+                            ColorSpinorField &Tmp2) const = 0;
 
     unsigned long long flops() const { return dirac->Flops(); }
 
-
     QudaMatPCType getMatPCType() const { return dirac->getMatPCType(); }
 
-    virtual int getStencilSteps() const = 0; 
+    virtual int getStencilSteps() const = 0;
 
     std::string Type() const { return typeid(*dirac).name(); }
-    
-    bool isStaggered() const {
+
+    /**
+       @brief return if the operator is a Wilson-type 4-d operator
+    */
+    bool isWilsonType() const
+    {
+      return (Type() == typeid(DiracWilson).name() || Type() == typeid(DiracWilsonPC).name()
+              || Type() == typeid(DiracClover).name() || Type() == typeid(DiracCloverPC).name()
+              || Type() == typeid(DiracCloverHasenbuschTwist).name()
+              || Type() == typeid(DiracCloverHasenbuschTwistPC).name() || Type() == typeid(DiracTwistedMass).name()
+              || Type() == typeid(DiracTwistedMassPC).name() || Type() == typeid(DiracTwistedClover).name()
+              || Type() == typeid(DiracTwistedCloverPC).name()) ?
+        true :
+        false;
```

Comment:
A non-blocking comment here: in the future if someone adds a new dirac type, but do not update the list here, things could go wrong. A possible solution is to have a pure virtual `isWilsonType()` for `Dirac`, and all derived classes of `Dirac` are forced to have a definition of this method.

Reply:
The same applies to `isStaggerd` and `isDwf` as well.

Reply:
@hummingtree that's a good suggestion---I'll file an issue about it and take care of it later this week.